[Export] Add runtime assert to non-strict export

BoyuanFeng opened the pull request:
This PR moves insert_deferred_runtime_asserts from dynamo to torch.fx.passes and uses it to add runtime assertion for non-strict export.

Differential Revision: D55944267

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang

avikchaudhuri commented on torch/export/_trace.py:
```diff
@@ -1067,6 +1068,11 @@ def forward(self, *args, **kwargs):
+        insert_deferred_runtime_asserts(
+            exported_program.graph_module,
+            fake_mode.shape_env,
+            "non_strict_exported_program",
```
What is the name argument supposed to mean? If it's just for printing, maybe add the module name here?

avikchaudhuri commented on torch/fx/passes/runtime_assert.py:
```diff
@@ -0,0 +1,233 @@
+                    # useless right now
+                    (
+                        res,
+                        f"Deferred runtime assertion failed {ra.expr}",
```
Nit: for the user, "deferred" doesn't mean much here. Something like "Runtime assertion failed: ..." would be fine.

Ideally I'd like to know what the values for the free symbols in `ra.expr` are. Does including `res` help make the error clearer? (At run time it's going to simply be `False` I think, which is not great...)

E.g., when `ra.expr` is `-u1 <= 0` is it possible to know what node corresponded to `u1` from the `sympy_interp` call above? That would be sweet because then you could replace `u1` with the node name corresponding to it.

avikchaudhuri commented on torch/fx/passes/runtime_assert.py:
```diff
@@ -0,0 +1,233 @@
+                            return None
+
+                    graph.call_function(
+                        torch._constrain_as_value,
```
We shouldn't call this any more. Why not `torch._check`?

avikchaudhuri commented on torch/fx/passes/runtime_assert.py:
```diff
@@ -0,0 +1,233 @@
+
+            defs = []
+
+            # For every new unbacked symbol, we need an fx.Node representing
```
Actually this seems to be doing what I'm asking. Maybe you should do `add_runtime_asserts` using this info?

tugsbayasgalan commented on torch/_dynamo/output_graph.py:
```diff
@@ -1345,203 +1346,6 @@ def remove_unused(node):
 
-    # TODO: this is a generic pass that should live outside of Dynamo
-    def insert_deferred_runtime_asserts(self, root, name) -> None:
-        """
```
I think this pass should live in AOTAutograd, not just in export. With this change, torch.compile wouldn't be running the asserts anymore. (cc: @ezyang )

tugsbayasgalan replied:
Oh ok makes sense.

angelayi commented on torch/fx/passes/runtime_assert.py:
```diff
@@ -0,0 +1,241 @@
+    tensor propagation, so we cannot conveniently insert them into the FX graph
+    when they occur.  Instead, we accumulate them in the ShapeEnv, and in this
+    pass insert them into the graph as proper tests.
+    #"""
```
```suggestion
    """
```

angelayi commented on torch/fx/passes/runtime_assert.py:
```diff
@@ -0,0 +1,241 @@
+graph_code_log = torch._logging.getArtifactLogger(__name__, "graph_code")
+
+
+def get_example_value_key(node: fx.Node) -> Optional[str]:
```
nit: why not just return the FakeTensor directly here?

angelayi commented on torch/fx/passes/runtime_assert.py:
```diff
@@ -0,0 +1,241 @@
+                            convert(vr.upper),
+                        ),
+                    )
+                    graph.call_function(torch._check, (ge_check,))
+                    graph.call_function(torch._check, (le_check,))
```
Should these be `torch.ops.aten._async_assert.msg(ge_check, "msg")`? Because if we're running the pass after tracing non-strict, then all ops inserted need to be OpOverloads

ezyang commented:
Hurry up and land this, I need to monkey around with the logic and it will merge conflict

BoyuanFeng committed:
[Export] Add runtime assert to non-strict export (#123681)

Summary:
This PR moves `insert_deferred_runtime_asserts` from dynamo to torch.fx.passes and uses it to add runtime assertion for non-strict export.

X-link: https://github.com/pytorch/benchmark/pull/2226

Test Plan: CI

Differential Revision: D55944267
